atomics/include/desul/atomics/Generic.hpp

```diff
@@ -636,8 +653,40 @@ DESUL_INLINE_FUNCTION void atomic_dec(T* const dest,
   return atomic_sub(dest, T(1), order, scope);
 }
 
-}  // namespace desul
+template <typename T,
+          class SuccessMemoryOrder,
+          class FailureMemoryOrder,
+          class MemoryScope>
+DESUL_INLINE_FUNCTION bool atomic_compare_exchange_strong(T* const dest,
+                                                          T& expected,
+                                                          T desired,
+                                                          SuccessMemoryOrder success,
+                                                          FailureMemoryOrder failure,
+                                                          MemoryScope scope) {
+  T const old = atomic_compare_exchange(dest, expected, desired, success, scope);
+  if (old != expected) {
+    atomic_store(&expected, old, failure, scope);
```

Comment:
@crtrott I know you said it should not be an atomic store but I don't understand how `failure` gets into the play without it

Reply:
the problem is that our low level atomic_compare_exchange doesn't expose success and failure mode. But failure mode generally is weaker than success mode anyway.